### PR TITLE
Plan pricing

### PR DESCRIPTION
**Pull request by slavcodev**
Extends https://github.com/Rebilly/RebillyAPI/pull/474/

### CODE REVIEW

File: spec/definitions/Plans/Plan.yaml [hunk 1]
```diff
@@ -37,57 +36,17 @@ properties:
...
-          type: number
-          format: double
-          example: 99.95
-        maxQuantity:
-          description: |
-            If the value is `null`, it's assumed to be infinite.  It starts at the end of the previous bracket.
-            If there is no previous bracket, it starts at 1.
-          type: integer
-          example: 1
+  pricing:
```
Comment (slavcodev) [inline spec/definitions/Plans/Plan.yaml]: alternative name might be `billing`

File: spec/definitions/Plans/Formulas/tiered.yaml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      brackets:
+        description: |
+          The price backets, along with the price formula,
+          is used to calculate the amount to charge for the product on this plan on the invoice.
+
+          The "tiered" example:
+
+          Price per apple | Max quantity | Description
+          ----------------|--------------|------------
+          $5              | 1            | 0 to 1
```
Comment (adamaltman) [inline spec/definitions/Plans/Formulas/tiered.yaml]: how does someone buy 0?

File: spec/definitions/Plans/Formulas/volume.yaml [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+      brackets:
+        description: |
+          The price backets, along with the price formula,
+          is used to calculate the amount to charge for the product on this plan on the invoice.
+
+          The "volume" example:
+
+          Price per apple | Max quantity | Description
+          ----------------|--------------|------------
+          $5              | 1            | 0 to 1
```
Comment (adamaltman) [inline spec/definitions/Plans/Formulas/volume.yaml]: how does someone buy 0?

File: spec/definitions/Plans/Formulas/tiered.yaml (hunk 2 above)
Comment (slavcodev) [reply spec/definitions/Plans/Formulas/tiered.yaml]: I copied that from other docs.

File: spec/definitions/Plans/Formulas/flat-rate.yaml [hunk 4]
```diff
@@ -0,0 +1,15 @@
+allOf:
+  - $ref: "#/definitions/PlanPriceFormula"
+  - type: object
+    required:
+      - price
+    properties:
+      price:
+        description: If the price is 0, it's free.
```
Comment (2e3s) [inline spec/definitions/Plans/Formulas/flat-rate.yaml]: Maybe add some note how it differs from fixed fee

File: spec/definitions/Plans/Formulas/stairstep.yaml [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+          ------------|--------------|------------
+          $5          | 1            | 1
+          $13         | 5            | 2 to 5
+          $30         | null         | 6 or more
+
+          If someone bought 1 apple, it would be $5.
+
+          If someone bought 2 apples, it would be $13. For 2 to 5 apples, the price is $13 in any case.
+
+          Notice the language here.
```
Comment (2e3s) [inline spec/definitions/Plans/Formulas/stairstep.yaml]: Probably excessive sentence

File: spec/definitions/Plans/Plan.yaml [hunk 6]
```diff
@@ -37,57 +35,17 @@ properties:
...
+    $ref: "#/definitions/PlanPriceFormula"
```
[the remark this answers is not in the document]
Comment (slavcodev) [reply spec/definitions/Plans/Plan.yaml]: It is. This is base spec for all formulas.

File: spec/definitions/Plans/PlanPriceFormula.yaml [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+    type: string
+    description: |
+      The price formula determines what algorithm is used to calculate the invoice price based on a few factors,
+      - the quantity in the subscription (which may be variable if usage pricing, otherwise determined when creating the subscription)
+      - the price brackets data
+
+      To determine which formula is correct, please see the price formula documentation.
+    enum:
+      - fixed-fee
+      - flat-rate
```
Comment (2e3s) [inline spec/definitions/Plans/PlanPriceFormula.yaml]: This looks like a tiered price with 1 bracket. Isn't it excessive?

Comment (2e3s) [reply spec/definitions/Plans/PlanPriceFormula.yaml]: And `fixed-fee` is `stairstep` with 1 bracket too

Comment (slavcodev) [reply spec/definitions/Plans/PlanPriceFormula.yaml]: Technically yes. But I think merchants who don't have tiered pricing, will be more clear with this formula, when is needed to specify price per unit only.

Comment (slavcodev) [reply spec/definitions/Plans/PlanPriceFormula.yaml]: > And fixed-fee is stairstep with 1 bracket too
I do not think so. For `fixed-fee` you can't specify `max quantity`.

Comment (2e3s) [reply spec/definitions/Plans/PlanPriceFormula.yaml]: > I do not think so. For fixed-fee you can't specify max quantity.

`fixed-fee` is stairstep with 1 bracket with maxQuantity = null.

Comment (slavcodev) [reply spec/definitions/Plans/PlanPriceFormula.yaml]: Technically yes. But this is tricky and source of possible confusions and wrong usage. Also we should know when merchant have such pricing, e.q. base on this information we can render different checkout page, different UI, different invoices.

**Pull request merged**